src/plenigo/internal/utils/JWT.php

```diff
@@ -0,0 +1,193 @@
+                JWT::_handleRegularError('Signature verification failed');
+            }
+        }
+        return $payload;
+    }
+
+    /**
+     * Converts and signs a PHP object or array into a JWT string.
+     *
+     * @param object|array $payload PHP object or array
+     * @param string       $key     The secret key
+     *
+     * @return string      A signed JWT
+     * @uses jsonEncode
+     * @uses urlsafeB64Encode
+     */
+    public static function encode($payload, $key) {
+        $header = array('alg' => 'HS256');
+
+        $segments = array();
+        $segments[] = JWT::urlsafeB64Encode(JWT::jsonEncode($header));
+        $segments[] = JWT::urlsafeB64Encode(JWT::jsonEncode($payload));
+        $signing_input = implode('.', $segments);
+
+        $signature = JWT::sign($signing_input, $key);
+        $segments[] = JWT::urlsafeB64Encode($signature);
+
+        return implode('.', $segments);
+    }
+
+    /**
+     * Sign a string with a given key and algorithm.
+     *
+     * @param string $msg    The message to sign
+     * @param string $key    The secret key
+     *
+     * @return string          An encrypted message
+     * @throws \DomainException Unsupported algorithm was specified
+     */
+    public static function sign($msg, $key) {
+        return hash_hmac('sha256', $msg, $key, true);
+    }
+
+    /**
+     * Decode a JSON string into a PHP object.
+     *
+     * @param string $input JSON string
+     *
+     * @return object          Object representation of JSON string
+     * @throws \Exception Provided string was invalid JSON
+     */
+    public static function jsonDecode($input) {
+        $obj = json_decode($input);
+        if (function_exists('json_last_error') && $errno = json_last_error()) {
+            JWT::_handleJsonError($errno);
+        } else if ($obj === null && $input !== 'null') {
+            JWT::_handleRegularError('Null result with non-null input');
+        }
+        return $obj;
+    }
+
+    /**
+     * Encode a PHP object into a JSON string.
+     *
+     * @param object|array $input A PHP object or array
+     *
+     * @return string          JSON representation of the PHP object or array
+     * @throws \Exception Provided object could not be encoded to valid JSON
+     */
+    public static function jsonEncode($input) {
+        $json = json_encode($input);
+        if (function_exists('json_last_error') && $errno = json_last_error()) {
+            JWT::_handleJsonError($errno);
+        } else if ($json === 'null' && $input !== null) {
+            JWT::_handleRegularError('Null result with non-null input');
+        }
+        return $json;
+    }
+
+    /**
+     * Decode a string with URL-safe Base64.
+     *
+     * @param string $input A Base64 encoded string
+     *
+     * @return string A decoded string
+     */
+    public static function urlsafeB64Decode($input) {
+        $remainder = strlen($input) % 4;
```

Comment:
Yes but perhaps we should move all of the functionality to the the test class itself. We are delivering code "for the future". Normally those code will never be used again.